pkg/csi/fuse/nodeserver.go

```diff
@@ -155,13 +164,172 @@ func (ns *nodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
+		return nil, errors.Wrapf(err, "NodeUnstageVolume: error when patching labels on node %s", ns.nodeId)
+	}
+
 	return &csi.NodeUnstageVolumeResponse{}, nil
 }
 
 func (ns *nodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+	// The lock is to ensure CSI plugin labels the node in correct order
+	ns.mutex.Lock()
+	defer ns.mutex.Unlock()
+
+	// 1. get dataset namespace and name by volume id
+	namespace, name, err := getNamespacedNameByVolumeId(ns.client, req.GetVolumeId())
+	if err != nil {
+		klog.Errorf("NodeStageVolume: can't get namespace and name by volume id %s: %v", req.GetVolumeId(), err)
+		return nil, errors.Wrapf(err, "NodeStageVolume: can't get namespace and name by volume id %s", req.GetVolumeId())
+	}
+
+	// 2. Label node
+	fuseLabelKey := common.LabelAnnotationFusePrefix + namespace + "-" + name
+	var labelsToModify common.LabelsToModify
+	labelsToModify.Add(fuseLabelKey, "true")
+
+	node, err := kubeclient.GetNode(ns.client, ns.nodeId)
+	if err != nil {
+		klog.Errorf("NodeStageVolume: can't get node %s: %v", ns.nodeId, err)
+		return nil, errors.Wrapf(err, "NodeStageVolume: can't get node %s", ns.nodeId)
+	}
+
+	_, err = utils.ChangeNodeLabelWithPatchMode(ns.client, node, labelsToModify)
+	if err != nil {
+		klog.Errorf("NodeStageVolume: error when patching labels on node %s: %v", ns.nodeId, err)
+		return nil, errors.Wrapf(err, "NodeStageVolume: error when patching labels on node %s", ns.nodeId)
+	}
+
+	fluidPath := req.GetVolumeContext()["fluid_path"]
+	mountType := req.GetVolumeContext()["mount_type"]
+
+	// checkMountReady checks the fuse mount path every 3 second for 30 seconds in total.
+	err = checkMountReady(fluidPath, mountType)
+	if err != nil {
+		return nil, errors.Errorf("fuse pod on node %s is not ready", ns.nodeId)
+	}
+
 	return &csi.NodeStageVolumeResponse{}, nil
 }
 
 func (ns *nodeServer) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeRequest) (*csi.NodeExpandVolumeResponse, error) {
 	return nil, status.Error(codes.Unimplemented, "")
 }
+
+func (ns *nodeServer) NodeGetCapabilities(ctx context.Context, req *csi.NodeGetCapabilitiesRequest) (*csi.NodeGetCapabilitiesResponse, error) {
+	glog.V(5).Infof("Using default NodeGetCapabilities")
+
+	return &csi.NodeGetCapabilitiesResponse{
+		Capabilities: []*csi.NodeServiceCapability{
+			{
+				Type: &csi.NodeServiceCapability_Rpc{
+					Rpc: &csi.NodeServiceCapability_RPC{
+						Type: csi.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME,
+					},
+				},
+			},
+		},
+	}, nil
+}
+
+func checkMountInUse(volumeName string) (bool, error) {
+	var inUse bool
+	glog.Infof("Try to check if the volume %s is being used", volumeName)
+	if volumeName == "" {
+		return inUse, errors.New("volumeName is not specified")
+	}
+
+	// TODO: refer to https://github.com/kubernetes-sigs/alibaba-cloud-csi-driver/blob/4fcb743220371de82d556ab0b67b08440b04a218/pkg/oss/utils.go#L72
```

Comment:
Could you add more logging for determining whether the volume is used?

Reply:
Thx. Added

Reply:
why using klog somewhere, uing glog somewhere uing logger somewhere?